src/kbmod/search/logging.h

```diff
@@ -0,0 +1,221 @@
+  // but I don't know it. Advice appreciated. Also switch statement? pattern
+  // matching? Uppercase/lowercase? string isn't a type? what year are we in?
+
+  // https://docs.python.org/3/library/logging.html#logging-levels
+  enum LogLevel {
+    DEBUG = 10,
+    INFO = 20,
+    WARNING = 30,
+    ERROR = 40,
+    CRITICAL = 50
+  };
+
+  static std::unordered_map<std::string, LogLevel> StringToLogLevel {
+    {"DEBUG",    LogLevel::DEBUG},
+    {"INFO",     LogLevel::INFO},
+    {"WARNING",  LogLevel::WARNING},
+    {"ERROR",    LogLevel::ERROR},
+    {"CRITICAL", LogLevel::CRITICAL}
+  };
+
+  static std::unordered_map<LogLevel, std::string> LogLevelToString {
+    {LogLevel::DEBUG,    "DEBUG"},
+    {LogLevel::INFO,     "INFO"},
+    {LogLevel::WARNING,  "WARNING"},
+    {LogLevel::ERROR,    "ERROR"},
+    {LogLevel::CRITICAL, "CRITICAL"}
+  };
+
+
+  // dictionary of strings are used to configure Logger behavior such as minimal
+  // log level allowed to print, time-format, log format etc.
+  typedef std::unordered_map<std::string, std::string> sdict;
+
+  // Logger is a base class that dispatches the logging mechanism (IO and
+  // formatting) via the virtual method `log`
+  class Logger {
+  public:
+    std::string name;
+    sdict config;
+    LogLevel level_threshold;
+
+    Logger(std::string logger_name) :
+      name(logger_name), config(), level_threshold{LogLevel::WARNING}
+    {}
+
+    Logger(std::string logger_name, sdict conf) :
+      name(logger_name), config(conf) {
+      level_threshold = StringToLogLevel[config["level"]];
+    }
+
+    virtual void log(std::string level, std::string msg) = 0;
+    void debug(std::string msg) { log("DEBUG", msg); }
+    void info(std::string msg) { log("INFO", msg); }
+    void warning(std::string msg) { log("WARNING", msg); }
+    void error(std::string msg) { log("ERROR", msg); }
+    void critical(std::string msg) { log("CRITICAL", msg); }
+  };
+
+
+#ifdef Py_PYTHON_H
+  // Passes the logging to the Python-side Logger. The py::handle is the
+  // pybind11 container holding the reference to the Python Logger. The actual
+  // logging format and mechanism is handled Python-side by the logging module,
+  // so just pass the message and level onward and let it be resolved in Python
+  class PyLogger : public Logger {
+  private:
+    py::handle pylogger;
+
+  public:
+    PyLogger(py::handle logger) :
+      Logger(logger.attr("name").cast<std::string>()),
+      pylogger(logger)
+    {}
+
+    virtual void log(std::string level, std::string msg){
+      for (char& ch : level) ch = std::tolower(ch);
+      pylogger.attr(level.c_str())(msg);
+    }
+  };
+#endif // Py_PYTHON_H
+
+
+  // Glorified std::cout. Unlike the PyLogger, CoutLogger is configurable.
+  // Configuration should contain `level`, `datefmt` and `format` keys:
+  // - `level`:  LogLevel enum value, the minimal level that is printed
+  // - `datefmt`: timestamp template usable with `strftime`
```

Comment:
In short - I can't enforce a format because I don't have a way to override Python logging module from C++ without providing a logging module. It's better to push the logging module from Python into C++. To satisfy this and alleviate your concerns I added more formatting options, really all that I have every actually used, so now even the non-Python loggers are fully configurable to match Python Loggers from a very pragmatic, practical view.

Long explanation:
If someone updates the base logging in Python - we have no "events", "hooks" or whatever to figure that out. So I would have to steal the whole logging management from Python and that's just hard. They have a very complicated logging configuration where `parent.child.grandchild` can all have different configs, and the `grandchild`, which uses stdout for example, propagates to `child`, which outputs logs into a file, which propagates to `parent` who sends an email. See [here](https://github.com/dirac-institute/trailblazer/blob/78abe53bc3f87f87213fec6e45f0c8fdc695708b/trail/config/logging.yaml#L4) for example. Once the Yaml/dict/file/toml config is read and parsed, these loggers are created, and these configs become various attributes on them or their anonymous proxy objects like `Formatter`, `Manager` etc. You can't even get the logging configuration you gave to Python, back as a dictionary from it. 
To enforce our setting from CPP - I'd have to run this whole tree and contextually somehow figure out where and what to set it to. 

Or I could just forget any of that exists and just dump everything to stdout. But then someone is going to want a file, someone is going to want a special record of errors and failures for large runs that don't require parsing GBs of logs etc. On top of that, because nobody is spending any time properly wrapping it, most of C++ code is in the position of reporting logs as a different module anyhow, like `StackSearch` should just report as `run_search`, `StampCreator` as the `CreateStamps` and so on. After that there's not a whole lot left, `StackSearch` more or less. We could just pre-declare a Python Logger with that name, via the config dict/file/yaml/toml, and invoke that from C++. The need to log from C++ as C++ doesn't seem very large at all. So I figured its better if I go the other way. Push Python stuff down to C++.